### PR TITLE
[shopsys] tokens are signed by private key

### PR DESCRIPTION
**Pull request by pesektomas**
| Q             | A
| ------------- | ---
|Description, reason for the PR| The tokens were signed using the Hmac key in the first implementation. We sign it with a private rsa key and verify it with its public key now.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/backend-api/install/build.xml.patch [hunk 1]
```diff
@@ -1,18 +1,14 @@
...
 +    <property name="path.backend-api" value="${path.vendor}/shopsys/backend-api"/>
      <property name="path.framework" value="${path.vendor}/shopsys/framework"/>
-
-     <property name="phpstan.level" value="4"/>
-
-     <import file="${path.framework}/build.xml"/>
+     <property name="path.frontend-api" value="${path.vendor}/shopsys/frontend-api"/>
 +    <import file="${path.backend-api}/build.xml"/>
 +
 +    <target name="composer-dev" depends="shopsys_framework.composer-dev,backend-api-oauth-keys-generate" description="Installs dependencies for development and generate OAuth keys."/>
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/backend-api/install/build.xml.patch:10]: Because I couldn't run the tests, because I couldn't apply the patch.

File: docs/frontend-api/authentication.md [hunk 2]
```diff
@@ -15,6 +15,11 @@ The access token has a short lifetime (15 minutes by default).
 In order to not force users to log in every 15 minutes, we have implemented refresh tokens that have a longer lifetime (14 days in the base).
 You can read more about refresh tokens [here](https://auth0.com/docs/tokens/concepts/refresh-tokens).
 
+Tokens are signed with a private key and verified with its public key in Shopsys framework.
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply docs/frontend-api/authentication.md]: 19c7c3c4b1ca6bf6f2275cf5ffe5e760bb293855

File: packages/frontend-api/build.xml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        <exec executable="${path.openssl.executable}" checkreturn="true">
+            <arg value="genrsa"/>
+            <arg value="-out"/>
+            <arg value="${path.config}/fe-api/private.key"/>
+            <arg value="2048"/>
+        </exec>
+        <exec executable="${path.openssl.executable}" checkreturn="true">
+            <arg value="rsa"/>
+            <arg value="-in"/>
+            <arg value="${path.config}/fe-api/private.key"/>
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/frontend-api/build.xml]: 887b934b5fe0ab236c9c7f19ec44f423ead7684c

File: packages/frontend-api/build.xml [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project name="frontend_api">
+    <property name="path.openssl.executable" value="openssl"/>
+    <property name="path.defuse-key-generate.executable" value="${path.bin}/generate-defuse-key"/>
+
+    <target name="frontend-api-generate-new-keys" description="Enable frontend API for all domains.">
```
Comment (grossmannmartin) [inline packages/frontend-api/build.xml:6]: Shouldn't be this target under production protection?

File: packages/frontend-api/build.xml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            <arg value="${path.config}/frontend-api/public.key"/>
+        </exec>
+        <exec executable="${path.defuse-key-generate.executable}" checkreturn="true" outputProperty="encryption_key" error="/dev/null">
+        </exec>
+
+        <echo file="${path.config}/frontend-api/.encryption_key" append="false">${encryption_key}</echo>
+    </target>
+
+    <target name="frontend-api-enable" depends="frontend-api-generate-new-keys" description="Enable frontend API for all domains.">
+        <copy file="${path.config}/packages/frontend_api.yaml.dist" tofile="${path.config}/packages/frontend_api.yaml" overwrite="true">
```
Comment (grossmannmartin) [inline packages/frontend-api/build.xml:28]: isn't possible to put the `frontend_api.yaml.dist` file into frontend-api package?

File: packages/frontend-api/build.xml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        </copy>
+    </target>
+
+    <target name="frontend-api-disable" description="Disable frontend API for all domains.">
+        <delete file="${path.config}/frontend-api/.encryption_key" />
```
Comment (grossmannmartin) [inline packages/frontend-api/build.xml:33]: Can I ask why is this file necessary (`.encryption_key`)?

File: packages/frontend-api/build.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/frontend-api/build.xml]: 14c1d0a120ff6f7f9fcf679f338a97dc1237dea0

File: docs/installation/installation-using-docker-on-production-server.md [hunk 7]
```diff
@@ -368,6 +368,11 @@ With login `superadmin` and password `admin123` we can do it via these urls:
 
 Now we go to `/admin/dashboard/` and fulfill all requests that are demanding for us by red colored links.
 
+If you want to use the frontend API, you have to generate a pair of keys at the first time you deploy. 
+You can do this with `./phing frontend-api-enable` target or, if `packages/frontend_api.yaml` configuration file exist, with `./phing frontend-api-generate-new-keys` target. 
+The keys are stored in the `config/frontend-api/` folder.
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply docs/installation/installation-using-docker-on-production-server.md:373]: d2ad5e7f9fdfe7a472e062c993bdac625728eba3

File: packages/frontend-api/build.xml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            <arg value="${path.config}/frontend-api/private.key"/>
+            <arg value="2048"/>
+        </exec>
+        <exec executable="${path.openssl.executable}" checkreturn="true">
+            <arg value="rsa"/>
+            <arg value="-in"/>
+            <arg value="${path.config}/frontend-api/private.key"/>
+            <arg value="-pubout"/>
+            <arg value="-out"/>
+            <arg value="${path.config}/frontend-api/public.key"/>
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/frontend-api/build.xml:19]: 82a027dbb469c96a2102fa7b7d890c4ea5238125

File: packages/frontend-api/build.xml (hunk 4 above)
Comment (pesektomas) [reply packages/frontend-api/build.xml]: 28b24330b4eaeebcc2d1ffce03d4d0f5a448cbab

File: packages/frontend-api/build.xml (hunk 5 above)
Comment (pesektomas) [reply packages/frontend-api/build.xml]: 28b24330b4eaeebcc2d1ffce03d4d0f5a448cbab

File: packages/frontend-api/build.xml (hunk 6 above)
Comment (pesektomas) [reply packages/frontend-api/build.xml:33]: Really I don't know. I suppose for some extra key operations, such as recovering a lost public key?

Comment (grossmannmartin) [reply packages/frontend-api/build.xml:33]: From what I found, `generate-defuse-key` is part of https://github.com/defuse/php-encryption package and it's script which generates a random encryption key. 
I don't think it's somehow related to generated keys with openssl.
I tried to remove the file and tokens were generated and used properly. 

If we don't have a use-case for this file, I suggest to remove it as it may confuse people.

File: packages/frontend-api/build.xml (hunk 5 above)
Comment (grossmannmartin) [reply packages/frontend-api/build.xml]: thanks for the fix (actually the commit is 91448745b9cbc1c392e7f16d56ddb4549dc1ea58 ;))

```
<copy file="${path.packages}/frontend-api/src/Resources/config/frontend_api.yaml.dist" tofile="${path.config}/packages/frontend_api.yaml" overwrite="true">
```

this will work only in monorepo, `path.packages` is not available in plain project-base.
I think 
```
<copy file="${path.frontend-api}/src/Resources/config/frontend_api.yaml.dist" tofile="${path.config}/packages/frontend_api.yaml" overwrite="true">
```
could be used instead

Comment (pesektomas) [reply packages/frontend-api/build.xml]: Sorry for bad commit hash. Fix is in this (perhaps correct) commit 2a65f09bd6b7df3913837a665f091bc3a4f32cfa

File: packages/frontend-api/build.xml (hunk 6 above)
Comment (pesektomas) [reply packages/frontend-api/build.xml]: ok, I removed it 2a65f09bd6b7df3913837a665f091bc3a4f32cfa
